### ISSUE TITLE
Get rid of `do_queue`?

### ISSUE THREAD

**Issue by cgogolin** [XanaduAI/pennylane#161]
The `Operation` class has in its `__init__()` the slightly obscure `do_queue=True` parameter. It not only clutters the docs (unless manual docstrings are given) but it is also not very well documented and the feature should be irrelevant (and thus hidden, imo) from the user. 

As far as I can tell, this option is actually set to `False` only in two places:

The first one, in `default_gaussian.py` is actually even marked as a hack:
```
    # HACK, we need access to the Poly instance in order to expand the matrix!
    op = qml.expval.PolyXP(Q, wires=wires, do_queue=False)
    Q = op.heisenberg_obs(N)
```
Only `Q` is used in the following code, so it would be better to obtain `Q` here without having to first instanciate an `PolyXP()` object. In a similar situation it was recently argued that we should accept code duplication in order to keep the plugin code self-contained. I would argue that the same rational could/should be applied here.

The second is in `qnode.py` in `_pd_analytic()`. Here, `PolyXP()` is called to compute the observable to measure via the "2nd order method" and `do_queue=False` is used to prevent the computed observables from adding themselves to the queue of the `qnode`, as they are only to be passed on to the device via `execute()`.

In conclusion, I would suggest to remove the `do_queue` property from `Operation`, create a private sub-class of `PolyXP` with an overwritten `queue()` method that does nothing and use that instead of the regular, user facing, `PolyXP` in these to places. To make the plugin code fully self-contained, one could further duplicate the code from `PolyXP.heisenberg_obs()` in the `default_gaussian` plugin.

Here all appearances of `do_queue` in `/pennylane/`:
```
grep --color -nH -r -e  'do_queue' * 
operation.py:190:        do_queue (bool): Indicates whether the operation should be
operation.py:253:    def __init__(self, *args, wires=None, do_queue=True):
operation.py:300:        if do_queue:
operation.py:433:        do_queue (bool): Indicates whether the operation should be immediately
Binary file plugins/__pycache__/default_gaussian.cpython-35.pyc matches
plugins/default_gaussian.py:636:    op = qml.expval.PolyXP(Q, wires=wires, do_queue=False)
Binary file __pycache__/qnode.cpython-35.pyc matches
Binary file __pycache__/operation.cpython-35.pyc matches
qnode.py:767:                    return pennylane.expval.PolyXP(qp, wires=range(w), do_queue=False)
```

**Comment by josh146** [comment]
> The Operation class has in its __init__() the slightly obscure do_queue=True parameter. It not only clutters the docs (unless manual docstrings are given) but it is also not very well documented and the feature should be irrelevant (and thus hidden, imo) from the user.

The history of `do_queue` is that it was added for the purposes of being able to unit test the base `Operation` class, as well as all inherited operations in the `ops` module, without the added overhead of interfacing through a QNode.

I'm undecided what the best approach is for the unit tests - some way of instantiating the operations outside of QNodes are needed for good atomic unit tests, but using `unittest.mock` everywhere has the potential to make the unit tests hard to read and maintain.

> Only Q is used in the following code, so it would be better to obtain Q here without having to first instanciate an PolyXP() object. In a similar situation it was recently argued that we should accept code duplication in order to keep the plugin code self-contained. I would argue that the same rational could/should be applied here.

Agree 100%.

> The second is in qnode.py in _pd_analytic(). Here, PolyXP() is called to compute the observable to measure via the "2nd order method" and do_queue=False is used to prevent the computed observables from adding themselves to the queue of the qnode, as they are only to be passed on to the device via execute().

Agree

**Comment by johannesjmeyer** [comment]
I'm in favour of patching `Operation.queue` in this case. I think it is suboptimal to have parameters that exist purely for test-purposes in the codebase, especially when it is easily achievable otherwise.

I don't know exactly how to do it, but I'm pretty sure that one could implement a `suppress_queuing` fixture or, even better, a context manager that supresses queuing in its context by patching `Operation.queue`.

**Comment by josh146** [comment]
@johannesjmeyer, since last discussion on this issue, we have noticed use cases for `do_queue` outside of unit tests. For example, there are some optimizers that may need to modify the QNode queue dynamically.

One solution is to remove the exception that is raised if an operation is used outside of a QNode.

Queueing now becomes implicit if inside a QNode, and the operation does nothing outside of a QNode (which I think makes sense)

**Comment by johannesjmeyer** [comment]
> @johannesjmeyer, since last discussion on this issue, we have noticed use cases for `do_queue` outside of unit tests. For example, there are some optimizers that may need to modify the QNode queue dynamically.
> 
Could you point me to an example?

> Queueing now becomes implicit if inside a QNode, and the operation does nothing outside of a QNode (which I think makes sense)

Yeah, that definitely makes sense!

**Comment by quantshah** [comment]
> Queueing now becomes implicit if inside a QNode, and the operation does nothing outside of a QNode (which I think makes sense)

If we can make sure that the queuing happens only when called inside the QNode and not inside `expval(PauliX(0))` then we can get rid of the need to `QNode._current_context.queue.remove(ops)` inside the `expval.__call__` method. I do not know if there can be a selective check like that to stop queuing if called inside the `expval` function.

Also makes sense to make it implicit.

**Comment by co9olguy** [comment]
Suggested simple fix:

- If an op is created outside of context managers, then it is not added to any `queue` (similar to what happens when `do_queue=False`)
- If an op is created inside a relevant context manager, it is automatically added in the corresponding circuit queue (similar to current default behaviour)
- If an op is created inside a relevant context manager, but the user does not want to add it to the queue, they specify this with an extra keyword argument, e.g., `_queue=False`

**Comment by josh146** [comment]
In addition, it feels hacky to pull ops out of the queue just to update return_type, then put them right back. There are three potential solutions, which differ in scope:

* Re-think the queuing behaviour across the board (probably too hard)

* Remove queuing from just observables (since it is not needed), and have the expval/sample/var functions perform the queuing

* See if it's possible to find and update the observable in the queue, without removing it?

I'm thinking option number (2) is preferable.

**Comment by co9olguy** [comment]
Closed by #359
